feat: retrieving user's purchased games

alekpentchev opened the pull request:
I think I'm ready to submit my first PR to this repo. And because of that I'm quite sure there might be some things worth changing. So here I'm counting on your support :) 
Few notes I want to make:

- I tried to follow the conventions from other existing functions
- I have extracted a lot of the GraphQL query variables to be available as params, e.g. size of the returned games list etc. This is not only for performance reasons if someone wouldn't like to fetch hundreds of games at once
- The URL the code generates seems to be working properly
- I had a problem with finding meaningful values for the `conceptId` property. For every game I got a null value. That's why for now I've added type definition as `null | unknown`
- I'm not sure if having base URLs in separate files is the best approach. What do you think about keeping them in one file which is used to import the URLs from?

Feel free to comment everything. I'm open for every discussion and recommendation.

wescopeland commented:
Thank you for submitting a PR! I will not have time to do a comprehensive review until Monday, but rest assured this will not get buried.

Just skimming over the PR, we will also need:
* Updates to the documentation website. To start the docs site:

```
cd website
yarn
yarn serve
```

* An update to _README.md_ that points to the new documentation page.

I haven't dug in to why CI is not properly running on GitHub Actions. Before I review on Monday, be sure to double check that `yarn verify` succeeds.

wescopeland commented:
Hi @alekpentchev ! Before I continue with a review, I want to double check that this work is not a duplicate of https://github.com/achievements-app/psn-api/pull/134

At first glance, this seems very similar.

alekpentchev commented:
> Hi @alekpentchev ! Before I continue with a review, I want to double check that this work is not a duplicate of #134
> 
> At first glance, this seems very similar.

Yeah, it looks similar. But both these operations serve a bit different purpose. The getUserGameList query shows all games the user has interacted with (not necessarily purchased), while getPurchasedGameList shows only games the user has actually bought or downloaded.

I wasn't aware that the library has this getUserGameList query being used. So probably I should rename my request to not confuse the users.

alekpentchev commented:
Added one more commit that changes the naming of my vars, objects etc. to avoid further confusion. Earlier I didn't notice that there's a similar looking function for getting recently played games.

One more difference I noticed is that for the `getPurchasedGameList` operation there's a bit more we get as a response than in the `getUserGameList` (used to fetch recently played games).

Please let me know if we're ok with it and if so I'll adjust the docs for this feature.

wescopeland commented on src/user/USER_GAMES_BASE_URL.ts:
```diff
@@ -0,0 +1,2 @@
+export const USER_GAMES_BASE_URL =
+  "https://web.np.playstation.com/api/graphql/v1/op";
```
This is duplicative of *GRAPHQL_BASE_URL.ts* in _src/graphql_.

I believe all the new code being added should be moved there so the module boundaries remain well-defined. This *USER_GAMES_BASE_URL.ts* file would ultimately be deleted, and the new function can reuse `GRAPHQL_BASE_URL`.

alekpentchev replied:
Good catch, thank you! This file was deleted and I'm now importing the `GRAPHQL_BASE_URL`

wescopeland commented on src/user/getUserPurchasedGames.ts:
```diff
@@ -0,0 +1,63 @@
+ * @param options.isActive Whether to return active games only. Defaults to `true`.
+ * @param options.subscriptionService The subscription service to filter by. Defaults to `"NONE"`.
+ */
+export const getUserPurchasedGames = async (
```
This is not currently being exported by its module. Consumers should be able to do a clean import like:

```ts
import { foo } from "psn-api";
```

However, I don't think that will work here because this function is not being exported by the module's _index.ts_ file.

Some consumers are doing:

```
import * as psnApi from "psn-api";
```

I think this import syntax may actually work for the new function, but I think this is probably an anti-pattern.

I have merged the latest `main` branch into this PR. It contains updates allowing you to run `yarn playground` for experimentation. Inside *src/__playground.ts* you'll find an import from the public-facing _index.ts_ file. We should be able to import all functions that are public-facing there.

alekpentchev replied:
Thank you for the comment! I noticed that GraphQL queries are located in a different folder so I moved the `getPurchasedGames` to the `graphql` folder in order to co-locate the code.
I have also added my function to the module's `index.ts` file.

Thanks to that I was able to successfully import and work with the code I have developed.

wescopeland commented:
> Added one more commit that changes the naming of my vars, objects etc. to avoid further confusion. Earlier I didn't notice that there's a similar looking function for getting recently played games.
> 
> One more difference I noticed is that for the `getPurchasedGameList` operation there's a bit more we get as a response than in the `getUserGameList` (used to fetch recently played games).
> 
> Please let me know if we're ok with it and if so I'll adjust the docs for this feature.

No problem, we will work together to make this the best it can be. I've left some initial feedback and we'll iterate further.

alekpentchev commented:
Added some changes based on the comments:

- reused `GRAPHQL_BASE_URL` instead of duplicating this variable
- registered the function in the module's `index.ts` file
- double checked if it works in the `__playground.ts` file
- moved my functions to the `graphql` folder in order to co-locate the GraphQL-related code
- changed functions names from `getUserPurchasedGames` to `getPurchasedGames` to be clearer when it comes to its meaning

wescopeland commented on src/graphql/getPurchasedGames.ts:
```diff
@@ -0,0 +1,68 @@
+export const getPurchasedGames = async (
+  authorization: AuthorizationPayload,
+  options: Partial<GetPurchasedGamesOptions> = defaultOptions
+): Promise<any> => {
```
This strips all type metadata from the return value. In other words, while the value is typed _within_ psn-api, it will be untyped for anyone consuming the library. You can see this in the playground:

```ts
  const purchasedGames = await getPurchasedGames(authorization);

  console.log(purchasedGames.data); // the type of `data` is `any` 💀
```

wescopeland commented on src/models/purchased-games-response.model.ts:
I do still wonder how much of this is actually duplicative. I just tried an experiment. In the playground file, I did:

```ts
  const recentlyPlayedGamesResponse = await getRecentlyPlayedGames(
    authorization
  );

  const purchasedGames = await getPurchasedGames(authorization);

  console.log(
    recentlyPlayedGamesResponse.data.gameLibraryTitlesRetrieve.games[0]
  );

  console.log(purchasedGames.data.purchasedTitlesRetrieve.games[0]);
```

The output is:

```ts
{
  __typename: 'GameLibraryTitle',
  conceptId: '231756',
  entitlementId: 'UP1018-PPSA01617_00-00MORTALKOMBAT11',
  image: {
    __typename: 'Media',
    url: 'https://agst.prod.dl.playstation.net/sgst/prod/00/PPSA01617_00/app/info/202/fi_1c056de0e68f29f2464d3d83e09645c42b1333f12316eb3d5fb5ba68c79fe926/icon0.png'
  },
  isActive: true,
  lastPlayedDateTime: '2023-02-07T19:13:46.020000Z',
  name: 'Mortal Kombat 11',
  platform: 'PS5',
  productId: 'UP1018-PPSA01617_00-00MORTALKOMBAT11',
  subscriptionService: 'NONE',
  titleId: 'PPSA01617_00'
}

{
  __typename: 'GameLibraryTitle',
  conceptId: null,
  entitlementId: 'UP0177-CUSA30446_00-SONICORIGINS0000',
  image: {
    __typename: 'Media',
    url: 'https://image.api.playstation.com/gs2-sec/appkgo/prod/CUSA30446_00/4/i_83f1fef9ce13d2b94907a21b304aefc6b9629129dc7504fb73ecf93d27a496fb/i/icon0.png'
  },
  isActive: true,
  isDownloadable: true,
  isPreOrder: false,
  name: 'Sonic Origins',
  platform: 'PS4',
  productId: 'UP0177-PPSA05324_00-SONICORIGINS0000',
  subscriptionService: 'NONE',
  titleId: 'CUSA30446_00'
}
```

With the exception of 3(?) props, the entities in the `games` arrays seem to have identical types.

What I think we should consider:
- We have a type called `RecentlyPlayedGame` in _recent-played-games-response.model.ts_.
- I think the common properties between the two responses should be "lifted up" into a single type. `GraphQLGameEntity` maybe?
- `RecentlyPlayedGame` can be refactored to extend the new type, like:

```ts
export type RecentlyPlayedGame = GraphQLGameEntity & {
  /** Example: Adding the unique properties in this call. */
  foo: string;
  bar: string;
  baz: string;
}
```

- `PurchasedGame` can do something similar:

```ts
export type PurchasedGame = GraphQLGameEntity & {
  foo: string;
  bar: string;
  baz: string;
}
```

Now we have great code re-use, great documentation re-use, and it's more clear which call tacks on which special properties to the base entity.

wescopeland commented on src/graphql/getPurchasedGames.ts:
```diff
@@ -0,0 +1,68 @@
+
+  return response;
+
+  return response;
```
```suggestion
```

nit

wescopeland commented on src/graphql/getPurchasedGames.ts:
```diff
@@ -0,0 +1,68 @@
+  authorization: AuthorizationPayload,
+  options: Partial<GetPurchasedGamesOptions> = defaultOptions
+): Promise<any> => {
+  const operationName = "getPurchasedGameList";
+  const variables = encodeURIComponent(JSON.stringify(options));
+  const extensions = encodeURIComponent(
+    '{"persistedQuery":{"version":1,"sha256Hash":"2c045408b0a4d0264bb5a3edfed4efd49fb4749cf8d216be9043768adff905e2"}}'
+  );
+  const url = `${GRAPHQL_BASE_URL}?operationName=${operationName}&variables=${variables}&extensions=${extensions}`;
```
Take a look at the implementation of _getUserRecentlyPlayedGames.ts_. I think the implementation of this new function should mirror it almost exactly, including moving the `sha256Hash` to a separate file.

wescopeland commented on src/graphql/getPurchasedGames.test.ts:
```diff
@@ -0,0 +1,107 @@
+    );
+
+    // ACT
+    const response = await getPurchasedGames(mockAuthorization);
+    // ASSERT
```
```suggestion
    const response = await getPurchasedGames(mockAuthorization);
    
    // ASSERT
```

nit

wescopeland commented:
Hi! This PR has exceeded 30 days of inactivity. I will be closing it as a courtesy. Please feel free to re-open it at any time you would like to revisit this work.
